### PR TITLE
feat: fetch arbitrary state-trees

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:
- New command: `forest-cli state fetch {CID}`
- The command uses libp2p-bitswap to fetch an IPLD graph in parallel and store it in the database.

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: node/rpc/src/state_api.rs [hunk 1]
```diff
@@ -186,3 +190,91 @@ pub(crate) async fn state_wait_msg<DB: Blockstore + Clone + Send + Sync + 'stati
...
+    task_set.spawn(async move { Ok(Ipld::Link(root_cid)) });
+
+    // Iterate until there are no more ipld nodes to traverse
+    while let Some(result) = task_set.join_next().await {
+        match result? {
+            Ok(ipld) => {
+                for new_cid in ipld.iter().filter_map(&mut get_ipld_link) {
+                    counter += 1;
+                    if counter % 1_000 == 0 {
+                        // set RUST_LOG=forest_rpc::state_api=debug to enable these printouts.
```
Comment (hanabi1224) [inline node/rpc/src/state_api.rs:235]: Maybe leave a TODO and refer to https://github.com/ChainSafe/forest/issues/2955 if it's meant to be replaced by the desired progress bar

File: node/rpc/src/state_api.rs [hunk 2]
```diff
@@ -186,3 +190,91 @@ pub(crate) async fn state_wait_msg<DB: Blockstore + Clone + Send + Sync + 'stati
...
+                                let permit = sem.acquire_owned().await?;
+                                let (tx, rx) = flume::bounded(1);
+                                network_send
+                                    .send_async(NetworkMessage::BitswapRequest {
+                                        epoch: 0,
+                                        cid: new_cid,
+                                        response_channel: tx,
+                                    })
+                                    .await?;
+                                // Bitswap requests do not fail. They are just ignored if no-one has
```
Comment (hanabi1224) [inline node/rpc/src/state_api.rs:258]: FYI there is a `send_dont_have` option in `BitswapRequest` to ask peers to respond even if they don't have the requested block, if that could be beneficial here.

File: node/rpc/src/state_api.rs [hunk 3]
```diff
@@ -186,3 +190,91 @@ pub(crate) async fn state_wait_msg<DB: Blockstore + Clone + Send + Sync + 'stati
...
+/// consensus rules.
+pub(crate) async fn state_fetch_root<DB: Blockstore + Clone + Sync + Send + 'static, B: Beacon>(
+    data: Data<RPCState<DB, B>>,
+    Params((CidJson(root_cid),)): Params<StateFetchRootParams>,
+) -> Result<StateFetchRootResult, JsonRpcError> {
+    const MAX_CONCURRENT_REQUESTS: usize = 16;
+    const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+    let sem = Arc::new(Semaphore::new(MAX_CONCURRENT_REQUESTS));
+    let mut seen: HashSet<Cid> = HashSet::new();
```
Comment (hanabi1224) [inline node/rpc/src/state_api.rs:216]: nit: maybe `CidHashSet`  here if collision is not critical

Comment (lemmih) [reply node/rpc/src/state_api.rs:216]: I would use `CidHashSet` if `insert` didn't take an extra callback function. :)

File: node/rpc/src/state_api.rs (hunk 1 above)
Comment (lemmih) [reply node/rpc/src/state_api.rs:235]: Showing a progress bar requires knowing how much work is left. We should show some progress indicators in the client, though. Maybe as a separate PR.

**Comment by lemmih** [comment]
> 1. Can we add some tests for this feature? Both where the Cid we know is present and where we know it won't be there.

I added a test that fetches the state-root of epoch 1. It _should_ always be available but it's not 100% guaranteed to stay available in the future. If it becomes a problem, we can switch to walking the genesis block which we always have available.

> 2. Could this be used instead of downloading actor bundles from Github? Not in this PR, but just a thought. It would give us some resilience from GH outages (but also make us dependent on other peers).

Yes, we should be able to fetch the bundles. I'll leave it for another PR.

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**